Blocks/Persistence/file.h

```diff
@@ -131,38 +144,54 @@ class FilePersister {
     FilePersisterImpl& operator=(FilePersisterImpl&&) = delete;
 
     explicit FilePersisterImpl(const std::string& filename)
-        : filename(filename),
-          appender(filename, std::ofstream::app),
-          end(ValidateFileAndInitializeNext(filename, offset, timestamp)) {
+        : filename(filename), appender(filename, std::ofstream::app), head_offset(0) {
+      ValidateFileAndInitializeHead();
       if (appender.bad()) {
         CURRENT_THROW(PersistenceFileNotWritable(filename));
       }
     }
 
     // Replay the file but ignore its contents. Used to initialize `end` at startup.
-    static end_t ValidateFileAndInitializeNext(const std::string& filename,
-                                               std::vector<std::streampos>& offset,
-                                               std::vector<std::chrono::microseconds>& timestamp) {
+    void ValidateFileAndInitializeHead() {
       std::ifstream fi(filename);
       if (!fi.bad()) {
         // Read through all the lines.
         // Let `IteratorOverFileOfPersistedEntries` maintain its own `next_`, which later becomes `this->end`.
         // While reading the file, record the offset of each record and store it in `offset`.
         IteratorOverFileOfPersistedEntries<ENTRY> cit(fi, 0, 0);
         std::streampos current_offset(0);
-        while (cit.ProcessNextEntry([&fi, &offset, &timestamp, &current_offset](const idxts_t& current, const char*) {
-          CURRENT_ASSERT(current.index == offset.size());
-          CURRENT_ASSERT(current.index == timestamp.size());
-          offset.push_back(current_offset);
-          timestamp.push_back(current.us);
-          current_offset = fi.tellg();
-        })) {
+        auto head = std::chrono::microseconds(-1);
+        while (cit.ProcessNextEntry(
+            [&](const idxts_t& current, const char*) {
+              CURRENT_ASSERT(current.index == offset.size());
+              CURRENT_ASSERT(current.index == timestamp.size());
+              if (!(current.us > head)) {
+                CURRENT_THROW(ss::InconsistentTimestampException(head + std::chrono::microseconds(1), current.us));
+              }
+              offset.push_back(current_offset);
+              timestamp.push_back(current.us);
+              current_offset = fi.tellg();
+              head = current.us;
+              head_offset = 0;
+            },
+            [&](const char* key, const char* value) {
+              if (key == constants::kHeadDirective) {
```

Comment:
Yeah, and it also means that we need a proper test for this part of the code :-/